tests/functional/test_cache.py

```diff
@@ -0,0 +1,101 @@
+    #   without pretty much re-implementing the entire cache info command.
+    assert 'Packages: ' in result.stdout
+
+
+def test_cache_list(script, monkeypatch):
+    cache_dir = _cache_dir(script)
+    wheel_cache_dir = os.path.join(cache_dir, 'wheels')
+    destination = os.path.join(wheel_cache_dir, 'arbitrary', 'pathname')
+    os.makedirs(destination)
+    with open(os.path.join(destination, 'yyy-1.2.3.whl'), 'w'):
+        pass
+    with open(os.path.join(destination, 'zzz-4.5.6.whl'), 'w'):
+        pass
+    result = script.pip('cache', 'list')
+    assert 'yyy-1.2.3' in result.stdout
+    assert 'zzz-4.5.6' in result.stdout
+    shutil.rmtree(os.path.join(wheel_cache_dir, 'arbitrary'))
+
+
+def test_cache_list_too_many_args(script, monkeypatch):
+    script.pip('cache', 'list', 'aaa', 'bbb',
+               expect_error=True)
+
+
+def test_cache_list_with_pattern(script, monkeypatch):
+    cache_dir = _cache_dir(script)
+    wheel_cache_dir = os.path.join(cache_dir, 'wheels')
+    destination = os.path.join(wheel_cache_dir, 'arbitrary', 'pathname')
+    os.makedirs(destination)
+    with open(os.path.join(destination, 'yyy-1.2.3.whl'), 'w'):
+        pass
+    with open(os.path.join(destination, 'zzz-4.5.6.whl'), 'w'):
+        pass
+    result = script.pip('cache', 'list', 'zzz')
+    assert 'yyy-1.2.3' not in result.stdout
+    assert 'zzz-4.5.6' in result.stdout
+    shutil.rmtree(os.path.join(wheel_cache_dir, 'arbitrary'))
+
+
+def test_cache_remove(script, monkeypatch):
+    cache_dir = _cache_dir(script)
+    wheel_cache_dir = os.path.join(cache_dir, 'wheels')
+    destination = os.path.join(wheel_cache_dir, 'arbitrary', 'pathname')
+    os.makedirs(destination)
+    with open(os.path.join(wheel_cache_dir, 'yyy-1.2.3.whl'), 'w'):
+        pass
+    with open(os.path.join(wheel_cache_dir, 'zzz-4.5.6.whl'), 'w'):
+        pass
+
+    script.pip('cache', 'remove', expect_error=True)
+    result = script.pip('cache', 'remove', 'zzz', '--verbose')
+    assert 'yyy-1.2.3' not in result.stdout
+    assert 'zzz-4.5.6' in result.stdout
+    shutil.rmtree(os.path.join(wheel_cache_dir, 'arbitrary'))
+
+
+def test_cache_remove_too_many_args(script, monkeypatch):
+    script.pip('cache', 'remove', 'aaa', 'bbb',
+               expect_error=True)
+
+
+def test_cache_purge(script, monkeypatch):
+    cache_dir = _cache_dir(script)
+    wheel_cache_dir = os.path.join(cache_dir, 'wheels')
+    destination = os.path.join(wheel_cache_dir, 'arbitrary', 'pathname')
+    os.makedirs(destination)
+    with open(os.path.join(wheel_cache_dir, 'yyy-1.2.3.whl'), 'w'):
+        pass
+    with open(os.path.join(wheel_cache_dir, 'zzz-4.5.6.whl'), 'w'):
+        pass
+
+    result = script.pip('cache', 'purge', 'aaa', '--verbose',
+                        expect_error=True)
+    assert 'yyy-1.2.3' not in result.stdout
+    assert 'zzz-4.5.6' not in result.stdout
+
+    result = script.pip('cache', 'purge', '--verbose')
```

Comment:
I'd split these two calls out into separate tests; to isolate the checks from each other.

Reply:
As a note to myself, I think I want to split this into two tests which test that:

- `cache purge <something with no matches> --verbose` raises an error _and_ doesn't delete anything. (the second part is tested implicitly by the second call working, but it'd be nice to _explicitly_ test it.)
- `cache purge --verbose` does not raise an error, deletes everything. (at the moment it only tests the output, but really it should check both that and that the file exists.)

and I'd like to add another test:

- `cache purge <something with matches> --verbose` does not raise an error, deletes all matching things, and _doesn't_ delete anything else.

Reply:
Those sound like very good tests 🙂